pkg/pmem-csi-operator/controller/deployment/controller_driver.go

```diff
@@ -67,194 +67,504 @@ var AllObjectTypes = []schema.GroupVersionKind{
+		if o, err = d.redeployServiceAccount(r); err != nil {
+			return fmt.Errorf("failed to update service account: %v", err)
+		}
+		allObjects = append(allObjects, o)
+		if o, err = d.redeployControllerService(r); err != nil {
+			return fmt.Errorf("failed to update controller service: %v", err)
+		}
+		allObjects = append(allObjects, o)
+		if o, err = d.redeployMetricsService(r); err != nil {
+			return fmt.Errorf("failed to update controller service: %v", err)
+		}
+		allObjects = append(allObjects, o)
+		if o, err = d.redeployCSIDriver(r); err != nil {
+			return fmt.Errorf("failed to update CSI driver: %v", err)
+		}
+		allObjects = append(allObjects, o)
+		if o, err = d.redeployControllerDriver(r); err != nil {
+			return fmt.Errorf("failed to update controller driver: %v", err)
+		}
+		allObjects = append(allObjects, o)
+		if o, err = d.redeployNodeDriver(r); err != nil {
+			return fmt.Errorf("failed to update node driver: %v", err)
+		}
+		allObjects = append(allObjects, o)
+		return nil
+	}
 
-		if err != nil {
-			klog.Warningf("Incompatible change of deployment occurred: %v", err)
+	if err := redeployAll(); err != nil {
+		return err
+	}
+
+	klog.Infof("Deployed '%d' objects.", len(allObjects))
+	// FIXME(avalluri): Limit the obsolete object deletion either to only version upgrades
+	// or on operator restart.
+	if err := d.deleteObsoleteObjects(r, allObjects); err != nil {
+		return fmt.Errorf("Delete obsolete objects failed with error: %v", err)
+	}
+
+	return nil
+}
+
+// getSubObject retrieves the latest revision of given object type from the API server
+// And checks if that object is owned by the current deployment CR
+func (d *PmemCSIDriver) getSubObject(r *ReconcileDeployment, obj apiruntime.Object) error {
+	objMeta, err := meta.Accessor(obj)
+	if err != nil {
+		return fmt.Errorf("internal error %T: %v", obj, err)
+	}
+
+	klog.Infof("Get: %T/%s", obj, objMeta.GetName())
+	if err := r.Get(obj); err != nil {
+		if errors.IsNotFound(err) {
+			klog.Infof("Not found: %T/%s", obj, objMeta.GetName())
+			return nil
 		}
+		return err
+	}
+	ownerRef := d.GetOwnerReference()
+	if !isOwnedBy(objMeta, &ownerRef) {
+		return fmt.Errorf("'%s' of type %T is not owned by '%s'", objMeta.GetName(), obj, ownerRef.Name)
 	}
 
-	objects := []apiruntime.Object{}
+	return nil
+}
 
-	// Force update all deployment objects
-	if updateAll {
-		klog.Infof("Updating all objects for deployment %q", d.Name)
-		objs, err := d.getDeploymentObjects()
-		if err != nil {
-			return true, err
+// updateSubObject writes the object changes to the API server.
+func (d *PmemCSIDriver) updateSubObject(r *ReconcileDeployment, op *ObjectPatch) error {
+	return op.Apply(r.client, d.Spec.Labels)
+}
+
+type redeployObject struct {
+	object     func() apiruntime.Object
+	modify     func(apiruntime.Object)
+	postUpdate func(apiruntime.Object)
+}
+
+func (d *PmemCSIDriver) redeploy(r *ReconcileDeployment, ro redeployObject) (apiruntime.Object, error) {
```

Comment:
It would be useful to explain each step in this function to help someone who is not familiar with the code understand what's going on. Don't just paraphrase the function names, but try to think of additional information that may be useful.

Reply:
> It would be useful to explain each step in this function to help someone who is not familiar with the code understand what's going on.

Added the function description, hope that helps.